### PR TITLE
New: Add enforcer metadata in handshake

### PR DESCRIPTION
**Pull request by sibicramesh**
--> Add enforcer metadata in handshake between two enforcer.
--> The plan is to add a simple 32 bit named field in the JWT exchanged between two enforcers on the packet path
--> If compressed tag in enforcer1 is V1 and if compressed tag in enforcer2 is V2, we drop the packet at network syn
--> If encrypt is set in enforcer1 but disabled on the fly to enforcer2, we drop the packet at network syn/ack

https://github.com/aporeto-inc/aporeto/issues/466

```
//    0             1              2               3               4
//  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
//  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
//  |     D     |CT|E|               R (reserved)                   |
//  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
//  D  [0:5]  - Datapath version
//  CT [6,7]  - Compressed tag type
//  E  [8]    - Encryption enabled
//  R  [9:31] - Unused currently
```

### CODE REVIEW

File: controller/internal/enforcer/nfqdatapath/claimsheader.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+// parseClaimsHeaderAttr returns the claimsHeader attribute set
+func parseClaimsHeaderAttr(claimsHeader []byte, mask uint8) uint8 {
+
+	data := binary.LittleEndian.Uint16(claimsHeader)
+
+	return uint8(data) & mask
+}
+
+func encryptionAttr(encrypt bool) uint8 {
```
Comment (0xmchadha) [inline controller/internal/enforcer/nfqdatapath/claimsheader.go]: Missing a comment for this function.

**Review by 0xmchadha — DISMISSED** [review]

File: controller/constants/constants.go [hunk 2]
```diff
@@ -97,6 +97,52 @@ const (
 	CompressedTagLengthV2 int = 4
 )
 
+// CompressionTypeMask defines the compression mask.
```
Comment (satyamsi) [inline controller/constants/constants.go]: move to new package as discussed yesterday.

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go [hunk 3]
```diff
@@ -330,7 +330,7 @@ func (p *Proxy) StartClientAuthStateMachine(downIP net.IP, downPort int, downCon
 			if err := downConn.SetWriteDeadline(time.Now().Add(5 * time.Second)); err != nil {
 				return false, err
 			}
-			token, err := p.tokenaccessor.CreateSynPacketToken(puContext, &conn.Auth)
+			token, err := p.tokenaccessor.CreateSynPacketToken(puContext, &conn.Auth, []byte{})
```
Comment (satyamsi) [inline controller/internal/enforcer/applicationproxy/tcp/tcp.go]: What is the behavior with proxy ?

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go [hunk 4]
```diff
@@ -736,6 +759,14 @@ func (d *Datapath) processNetworkSynAckPacket(context *pucontext.PUContext, conn
 		return nil, nil, nil
 	}
 
+	// NOTE: For backward compatibility, remove this check later
```
Comment (satyamsi) [inline controller/internal/enforcer/nfqdatapath/datapath_tcp.go:740]: Nope. We upgrade. If anyone side wants encryption, we encrypt - we dont drop connection.

File: controller/pkg/claimsheader/bytes.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package claimsheader
```
Comment (satyamsi) [inline controller/pkg/claimsheader/bytes.go:1]: claimsheader should be a part of this package.

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go [hunk 6]
```diff
@@ -326,8 +328,14 @@ func (d *Datapath) processApplicationSynPacket(tcpPacket *packet.Packet, context
 	// Create TCP Option
 	tcpOptions := d.createTCPAuthenticationOption([]byte{})
 
+	// We now generate the claims header
+	compressionType := d.secrets.(*secrets.CompactPKI).Compressed.CompressionTypeToMask()
```
Comment (satyamsi) [inline controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: Why Mask comparision ? Why not type comparison ?

Comment (satyamsi) [reply controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: NewClaimsHeader should have required params. 
- Version 
- CompressionType 
- Encrypt

File: controller/pkg/claimsheader/types.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+type ClaimsHeader struct {
+	// CompressionType represents compressed tag mode attribute
+	compressionType constants.CompressionTypeMask
+	// Encrypt represents enryption enabled attribute
+	encrypt bool
+	// Handshake type represents handshake version
+	handshakeVersion uint8
+}
+
+// boolToUint8 converts bool to uint8
```
Comment (satyamsi) [inline controller/pkg/claimsheader/types.go:13]: why is this needed ?

File: controller/pkg/claimsheader/types.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+func boolToUint8(e bool) uint8 {
+
+	if !e {
+		return 0x00
+	}
+
+	return EncryptionEnabledMask
+}
+
+// uint8ToBool converts uint8 to bool
```
Comment (satyamsi) [inline controller/pkg/claimsheader/types.go]: why is this needed ?

**Review by satyamsi — CHANGES_REQUESTED** [review]

File: controller/pkg/claimsheader/bytes.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+	return &ClaimsHeader{
+		compressionType:  constants.CompressionTypeMask(c.extractHeaderAttribute(constants.CompressionTypeBitMask.ToUint8())),
+		encrypt:          uint8ToBool(c.extractHeaderAttribute(EncryptionEnabledMask)),
+		handshakeVersion: c.extractHeaderAttribute(HandshakeVersion),
+	}
+}
+
+// extractHeaderAttribute returns the attribute from bytes
+func (c HeaderBytes) extractHeaderAttribute(mask uint8) uint8 {
```
Comment (varks) [inline controller/pkg/claimsheader/bytes.go]: mask has to 32 bits and data needs to be extracted as uint32.

File: collector/interfaces.go [hunk 10]
```diff
@@ -36,6 +36,10 @@ const (
 	APIPolicyDrop = "api"
 	// UnableToDial indicates that the proxy cannot dial out the connection
 	UnableToDial = "dial"
+	// CompressedTagMismatch indicates that the compressed tag version is dissimilar
+	CompressedTagMismatch = "compressedtagmismatch"
+	// EncryptionMismatch indicates that the policy encryption varies between client and server enforcer
```
[the remark this answers is not in the document]
Comment (sibicramesh) [reply collector/interfaces.go]: This will be done in a separate PR

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go (hunk 4 above)
Comment (sibicramesh) [reply controller/internal/enforcer/nfqdatapath/datapath_tcp.go:740]: Yes, this will be done in next PR

File: controller/pkg/claimsheader/bytes.go (hunk 5 above)
Comment (sibicramesh) [reply controller/pkg/claimsheader/bytes.go:1]: this is a part of claimsheader

File: controller/pkg/claimsheader/types.go (hunk 7 above)
Comment (sibicramesh) [reply controller/pkg/claimsheader/types.go:13]: to convert the bool flag given to the bit thats needed to be set in bytes slice

File: controller/pkg/claimsheader/bytes.go (hunk 9 above)
Comment (sibicramesh) [reply controller/pkg/claimsheader/bytes.go]: Fixed

File: controller/pkg/claimsheader/types.go (hunk 8 above)
Comment (sibicramesh) [reply controller/pkg/claimsheader/types.go]: to convert the bit from the bytes to bool which will be populated in the struct for easy access

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go (hunk 3 above)
Comment (sibicramesh) [reply controller/internal/enforcer/applicationproxy/tcp/tcp.go]: Added

File: controller/constants/constants.go (hunk 2 above)
Comment (sibicramesh) [reply controller/constants/constants.go]: Moved

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "ff7ae15529bc77e2a043ab608f361699237de831"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "1da16edf258f936798b142ad9b4d26411426db9a"
  }
]
```

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go (hunk 6 above)
Comment (sibicramesh) [reply controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: It is better to have it optional because for example the tcp proxy doesn't need to have the datapath version, but as we discussed I've added both (version and compression) the fields in the complete datapath handshake

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "ff7ae15529bc77e2a043ab608f361699237de831"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "7873644a492aae284979809f419489526abe6220"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "ff7ae15529bc77e2a043ab608f361699237de831"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "958231f44d471db87ebd0ee5b87da4b3aaf8ab5a"
  }
]
```

File: controller/pkg/claimsheader/types.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+func boolToUint8(e bool) uint8 {
+
+	if !e {
+		return 0x00
+	}
+
+	return encryptionEnabledBit
+}
+
+// uint32ToBool converts uint8 to bool
```
Comment (0xmchadha) [inline controller/pkg/claimsheader/types.go:24]: Chose different names for these functions. uint32ToBool is too generic a name. IsEncryptionEnabled or something similar is better.

File: controller/pkg/secrets/compactpki.go [hunk 12]
```diff
@@ -115,7 +115,7 @@ func (p *CompactPKI) TransmittedKey() []byte {
 
 // AckSize returns the default size of an ACK packet
 func (p *CompactPKI) AckSize() uint32 {
-	return uint32(280)
+	return uint32(300)
```
Comment (0xmchadha) [inline controller/pkg/secrets/compactpki.go]: Create a constant here.

File: controller/pkg/claimsheader/types.go (hunk 11 above)
Comment (sibicramesh) [reply controller/pkg/claimsheader/types.go:24]: This is made to be generic. In future when we have another bool type field. We can pass the bitType to the function and return boolType. For example,
```
uint32ToBool(bit BitType)BoolType{
      swtich bit{
           case a:
                return x
           case b:
                return y
           .
           .
         }
}
```
(same goes for boolToUint8)

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "3cff6509cb5f5d9fa0624e9b770ef55249998f5a"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "33a86f89472a10535e9dab9afd910c699a35e7d9"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "039cd80545e62c74fd2437035b36aab448d586a9"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "541477e0fd05709ffc933299d08cb185b23d9fc8"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "039cd80545e62c74fd2437035b36aab448d586a9"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "541477e0fd05709ffc933299d08cb185b23d9fc8"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "039cd80545e62c74fd2437035b36aab448d586a9"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "541477e0fd05709ffc933299d08cb185b23d9fc8"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "15d4a50b9e18640b309c6851dfa1637401e98679"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "fe4e85a2fc4d1090a227f6ea94809fc56846be80"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "15d4a50b9e18640b309c6851dfa1637401e98679"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "08cf2b59d96ecf4478f0344fb8ed5bbd4bc2ea5b"
  }
]
```

**Review by 0xmchadha — DISMISSED** [review]

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go [hunk 13]
```diff
@@ -330,7 +334,11 @@ func (p *Proxy) StartClientAuthStateMachine(downIP net.IP, downPort int, downCon
 			if err := downConn.SetWriteDeadline(time.Now().Add(5 * time.Second)); err != nil {
 				return false, err
 			}
-			token, err := p.tokenaccessor.CreateSynPacketToken(puContext, &conn.Auth)
+			// We now generate the claims header
+			claimsHeaderBytes := claimsheader.NewClaimsHeader(
+				claimsheader.OptionCompressionType(p.secrets.(*secrets.CompactPKI).Compressed),
```
Comment (satyamsi) [inline controller/internal/enforcer/applicationproxy/tcp/tcp.go]: typecasting can fail ?

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go [hunk 14]
```diff
@@ -366,7 +385,10 @@ func (p *Proxy) StartClientAuthStateMachine(downIP net.IP, downPort int, downCon
 			if err := downConn.SetWriteDeadline(time.Now().Add(5 * time.Second)); err != nil {
 				return false, err
 			}
-			token, err := p.tokenaccessor.CreateAckPacketToken(puContext, &conn.Auth)
+			claimsHeaderBytes := claimsheader.NewClaimsHeader(
+				claimsheader.OptionCompressionType(p.secrets.(*secrets.CompactPKI).Compressed),
```
Comment (satyamsi) [inline controller/internal/enforcer/applicationproxy/tcp/tcp.go]: typecasting ?

File: controller/pkg/claimsheader/claimsheader.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+// ToBytes generates the 32-bit header in bytes
+//    0             1              2               3               4
+//  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//  |     D     |CT|E|               R (reserved)                   |
+//  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//  D  [0:5]  - Datapath version
+//  CT [6,7]  - Compressed tag type
+//  E  [8]    - Encryption enabled
+//  R  [9:31] - Unused currently
```
Comment (satyamsi) [inline controller/pkg/claimsheader/claimsheader.go]: Reserved

**Review by satyamsi — DISMISSED** [review]
Address typecasting

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "15d4a50b9e18640b309c6851dfa1637401e98679"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "e9d02473bba245087a59a33a72ae7b580c5f9ea4"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "15d4a50b9e18640b309c6851dfa1637401e98679"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "c60609e81815fa8f28118a39e51a836161939860"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "15d4a50b9e18640b309c6851dfa1637401e98679"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "b88cfb3956dc0995ce41bc2510bb3c0734d18593"
  }
]
```

**Review by 0xmchadha — DISMISSED** [review]

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go [hunk 16]
```diff
@@ -9,6 +9,7 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"reflect"
 	"strconv"
```
Comment (dstiliadis) [inline controller/internal/enforcer/applicationproxy/tcp/tcp.go]: let's not use reflect in the data path.

File: controller/internal/enforcer/applicationproxy/applicationproxy.go [hunk 17]
```diff
@@ -261,7 +261,7 @@ func (p *AppProxy) registerAndRun(ctx context.Context, puID string, ltype common
 		c := httpproxy.NewHTTPProxy(p.collector, puID, caPool, appproxy, proxyMarkInt, p.secrets, p.registry)
 		return c, c.RunNetworkServer(ctx, listener, encrypted)
 	default:
-		c := tcp.NewTCPProxy(p.tokenaccessor, p.collector, puID, p.registry, p.cert, caPool)
+		c := tcp.NewTCPProxy(p.tokenaccessor, p.collector, puID, p.registry, p.cert, caPool, p.secrets)
 		return c, c.RunNetworkServer(ctx, listener, encrypted)
```
Comment (dstiliadis) [inline controller/internal/enforcer/applicationproxy/applicationproxy.go]: Don't pass the secrets here. Do the calculation of the header you need and pass that. No reason to go over all the logic for every packet.

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go [hunk 18]
```diff
@@ -330,7 +335,17 @@ func (p *Proxy) StartClientAuthStateMachine(downIP net.IP, downPort int, downCon
...
+			compactPKI, ok := p.secrets.(*secrets.CompactPKI)
+			if !ok {
+				zap.L().Error("Secrets does not hold compactPKI type, so type assertion failed", zap.Reflect("type", reflect.TypeOf(p.secrets)))
+				return isEncrypted, secrets.ErrNotCompactPKI
+			}
+			claimsHeaderBytes := claimsheader.NewClaimsHeader(
+				claimsheader.OptionCompressionType(compactPKI.Compressed),
+			).ToBytes()
+			token, err := p.tokenaccessor.CreateSynPacketToken(puContext, &conn.Auth, claimsHeaderBytes)
 			if err != nil {
```
Comment (dstiliadis) [inline controller/internal/enforcer/applicationproxy/tcp/tcp.go]: This should be done once ahead of time and not on every request.

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go [hunk 19]
```diff
@@ -366,7 +392,15 @@ func (p *Proxy) StartClientAuthStateMachine(downIP net.IP, downPort int, downCon
 			if err := downConn.SetWriteDeadline(time.Now().Add(5 * time.Second)); err != nil {
 				return false, err
 			}
-			token, err := p.tokenaccessor.CreateAckPacketToken(puContext, &conn.Auth)
```
Comment (dstiliadis) [inline controller/internal/enforcer/applicationproxy/tcp/tcp.go]: same as above.

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go [hunk 20]
```diff
@@ -450,7 +490,16 @@ func (p *Proxy) StartServerAuthStateMachine(ip net.IP, backendport int, upConn n
...
+			compactPKI, ok := p.secrets.(*secrets.CompactPKI)
+			if !ok {
+				zap.L().Error("Secrets does not hold compactPKI type, so type assertion failed", zap.Reflect("type", reflect.TypeOf(p.secrets)))
+				return isEncrypted, secrets.ErrNotCompactPKI
+			}
+			claimsHeaderBytes := claimsheader.NewClaimsHeader(
+				claimsheader.OptionCompressionType(compactPKI.Compressed),
+				claimsheader.OptionEncrypt(conn.PacketFlowPolicy.Action.Encrypted()),
+			).ToBytes()
+			claims, err := p.tokenaccessor.CreateSynAckPacketToken(puContext, &conn.Auth, claimsHeaderBytes)
```
Comment (dstiliadis) [inline controller/internal/enforcer/applicationproxy/tcp/tcp.go]: same as previous comments. Calculate once.

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go [hunk 21]
```diff
@@ -4,14 +4,17 @@ package nfqdatapath
 import (
 	"errors"
 	"fmt"
+	"reflect"
 	"strconv"
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: no reflect on the data path even for logs

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go [hunk 22]
```diff
@@ -322,8 +325,19 @@ func (d *Datapath) processApplicationSynPacket(tcpPacket *packet.Packet, context
...
+	compactPKI, ok := d.secrets.(*secrets.CompactPKI)
+	if !ok {
+		zap.L().Error("Secrets does not hold compactPKI type, so type assertion failed", zap.Reflect("type", reflect.TypeOf(d.secrets)))
+		return nil, secrets.ErrNotCompactPKI
+	}
+	claimsHeaderBytes := claimsheader.NewClaimsHeader(
+		claimsheader.OptionCompressionType(compactPKI.Compressed),
+		claimsheader.OptionDatapathVersion(d.datapathVersion),
+	).ToBytes()
+
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: same as above. evaluate once.

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go [hunk 23]
```diff
@@ -384,7 +398,19 @@ func (d *Datapath) processApplicationSynAckPacket(tcpPacket *packet.Packet, cont
...
+		zap.L().Error("Secrets does not hold compactPKI type, so type assertion failed", zap.Reflect("type", reflect.TypeOf(d.secrets)))
+		return secrets.ErrNotCompactPKI
+	}
+	claimsHeaderBytes := claimsheader.NewClaimsHeader(
+		claimsheader.OptionCompressionType(compactPKI.Compressed),
+		claimsheader.OptionDatapathVersion(d.datapathVersion),
+		claimsheader.OptionEncrypt(conn.PacketFlowPolicy.Action.Encrypted()),
+	).ToBytes()
+
+	tcpData, err := d.tokenAccessor.CreateSynAckPacketToken(context, &conn.Auth, claimsHeaderBytes)
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: Too much cut and paste of code. All this needs to be calculated once and out of a function.

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go [hunk 24]
```diff
@@ -406,7 +432,16 @@ func (d *Datapath) processApplicationAckPacket(tcpPacket *packet.Packet, context
...
+		if !ok {
+			zap.L().Error("Secrets does not hold compactPKI type, so type assertion failed", zap.Reflect("type", reflect.TypeOf(d.secrets)))
+			return secrets.ErrNotCompactPKI
+		}
+		claimsHeaderBytes := claimsheader.NewClaimsHeader(
+			claimsheader.OptionCompressionType(compactPKI.Compressed),
+			claimsheader.OptionDatapathVersion(d.datapathVersion),
+		).ToBytes()
+		token, err := d.tokenAccessor.CreateAckPacketToken(context, &conn.Auth, claimsHeaderBytes)
 		if err != nil {
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: same here

File: controller/internal/enforcer/nfqdatapath/datapath_udp.go [hunk 25]
```diff
@@ -372,8 +375,19 @@ func (d *Datapath) processApplicationUDPSynPacket(udpPacket *packet.Packet, cont
...
+	if !ok {
+		zap.L().Error("Secrets does not hold compactPKI type, so type assertion failed", zap.Reflect("type", reflect.TypeOf(d.secrets)))
+		return secrets.ErrNotCompactPKI
+	}
+	claimsHeaderBytes := claimsheader.NewClaimsHeader(
+		claimsheader.OptionCompressionType(compactPKI.Compressed),
+		claimsheader.OptionDatapathVersion(d.datapathVersion),
+	).ToBytes()
+
 	udpOptions := d.CreateUDPAuthMarker(packet.UDPSynMask)
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/datapath_udp.go]: same here. all this code is cut and paste.

**Review by dstiliadis — CHANGES_REQUESTED** [review]
Needs refactoring and not cut-and-paste of the same section of code.

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go (hunk 21 above)
Comment (sibicramesh) [reply controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: removed

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go (hunk 18 above)
Comment (sibicramesh) [reply controller/internal/enforcer/applicationproxy/tcp/tcp.go]: > Needs refactoring and not cut-and-paste of the same section of code.

The claimsheader is not part of datapath, it is created to share information between two endpoints (for ex: appSYN to netSYN). To be more precise, I can't put create this claimsheader outside this function, 
```
	claimsHeaderBytes := claimsheader.NewClaimsHeader(
		claimsheader.OptionCompressionType(compactPKI.Compressed),
		claimsheader.OptionDatapathVersion(d.datapathVersion),
		claimsheader.OptionEncrypt(conn.PacketFlowPolicy.Action.Encrypted()),
	).ToBytes()
```

The third argument passed is the encryption state at that particular point. I can remove the other fields from the claimsheader but @satyamsi wants to have them in all headers.

**Comment by sibicramesh** [comment]
> Needs refactoring and not cut-and-paste of the same section of code.

The claimsheader is not part of datapath, it is created to share information between two endpoints (for ex: appSYN to netSYN). To be more precise, I can't create this claimsheader outside this function, 
```
	claimsHeaderBytes := claimsheader.NewClaimsHeader(
		claimsheader.OptionCompressionType(compactPKI.Compressed),
		claimsheader.OptionDatapathVersion(d.datapathVersion),
		claimsheader.OptionEncrypt(conn.PacketFlowPolicy.Action.Encrypted()),
	).ToBytes()
```

The third argument passed is the encryption state at that particular point. I can remove the other fields from the claimsheader but @satyamsi wants to have them in all headers.

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "15d4a50b9e18640b309c6851dfa1637401e98679"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "7ee4c7822b4390f8074be484f4a54c9549a54a17"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "aaa68c2eaccab5a9d53273e6780d79aaec61adef"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "7ee4c7822b4390f8074be484f4a54c9549a54a17"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "aaa68c2eaccab5a9d53273e6780d79aaec61adef"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "e2c5c40f3fd04b5111c763ce29f4234fc81757c7"
  }
]
```

File: controller/internal/enforcer/applicationproxy/applicationproxy.go (hunk 17 above)
Comment (sibicramesh) [reply controller/internal/enforcer/applicationproxy/applicationproxy.go]: fixed

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go (hunk 20 above)
Comment (sibicramesh) [reply controller/internal/enforcer/applicationproxy/tcp/tcp.go]: fixed

File: controller/internal/enforcer/nfqdatapath/datapath_udp.go (hunk 25 above)
Comment (sibicramesh) [reply controller/internal/enforcer/nfqdatapath/datapath_udp.go]: fixed

**Comment by sibicramesh** [comment]
--> Compressed type and datapath version has been moved to tokens and returns specific errors type, which is handled in datapath to report flow with proper drop reason
--> Encrypt stays in the datapath.
--> Extended the claimsheader pkg to set header fields and added unit tests for it

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "aaa68c2eaccab5a9d53273e6780d79aaec61adef"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "4c938a629c3ee776fc4969f3c5c9efdd17d708e1"
  }
]
```

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "aaa68c2eaccab5a9d53273e6780d79aaec61adef"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "c5106d129390b797d52245d8e92fa669ab7a71ca"
  }
]
```

**Review by dstiliadis — DISMISSED** [review]

**Comment by sibicramesh** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "enforcer-metadata",
    "component": "enforcerd",
    "pr-id": "980",
    "commit-sha": "aaa68c2eaccab5a9d53273e6780d79aaec61adef"
  },
  {
    "project": "enforcer-metadata",
    "component": "trireme-lib",
    "pr-id": "714",
    "commit-sha": "491ccac1edff62e4ca1654c272f86d888a8967af"
  }
]
```

**Review by dstiliadis — APPROVED** [review]
